x-pack/plugins/maps/public/classes/styles/vector/components/symbol/custom_icon_modal.tsx

```diff
@@ -0,0 +1,396 @@
+      defaultMessage: 'Select or drag and drop an SVG icon',
+    }),
+  getImageInputDescription: () =>
+    i18n.translate('xpack.maps.customIconModal.imageInputDescription', {
+      defaultMessage:
+        'SVGs without sharp corners and intricate details work best. Modifying the settings under Advanced options may improve rendering.',
+    }),
+  getInvalidFileLabel: () =>
+    i18n.translate('xpack.maps.customIconModal.invalidFileError', {
+      defaultMessage: 'Icon must be in SVG format. Other image types are not supported.',
+    }),
+  getNameInputLabel: () =>
+    i18n.translate('xpack.maps.customIconModal.nameInputLabel', {
+      defaultMessage: 'Name',
+    }),
+  getRadiusRangeLabel: () => (
+    <EuiToolTip
+      content={i18n.translate('xpack.maps.customIconModal.raduisRangeTooltip', {
+        defaultMessage:
+          'Adjusts the size of the signed distance function around the Alpha threshold as a percent of icon size.',
+      })}
+    >
+      <>
+        {i18n.translate('xpack.maps.customIconModal.radiusRangeLabel', {
+          defaultMessage: 'Radius',
+        })}{' '}
+        <EuiIcon color="subdued" type="questionInCircle" />
+      </>
+    </EuiToolTip>
+  ),
+  getResetButtonLabel: () =>
+    i18n.translate('xpack.maps.customIconModal.resetButtonLabel', {
+      defaultMessage: 'Reset',
+    }),
+  getSaveButtonLabel: () =>
+    i18n.translate('xpack.maps.customIconModal.saveButtonLabel', {
+      defaultMessage: 'Save',
+    }),
+};
+
+function getFileNameWithoutExt(fileName: string) {
+  const splits = fileName.split('.');
+  if (splits.length > 1) {
+    splits.pop();
+  }
+  return splits.join('.');
+}
+interface Props {
+  /**
+   * initial value for the id of image added to map
+   */
+  symbolId?: string;
+  /**
+   * initial value of the label of the custom element
+   */
+  label?: string;
+  /**
+   * initial value of the preview image of the custom element as a base64 dataurl
+   */
+  svg?: string;
+  /**
+   * intial value of alpha threshold for signed-distance field
+   */
+  cutoff: number;
+  /**
+   * intial value of radius for signed-distance field
+   */
+  radius: number;
+  /**
+   * title of the modal
+   */
+  title: string;
+  /**
+   * A click handler for the save button
+   */
+  onSave: (icon: CustomIcon) => void;
+  /**
+   * A click handler for the cancel button
+   */
+  onCancel: () => void;
+  /**
+   * A click handler for the delete button
+   */
+  onDelete?: (symbolId: string) => void;
+}
+
+interface State {
+  /**
+   * id of image added to map
+   */
+  symbolId: string;
+  /**
+   * label of the custom element to be saved
+   */
+  label: string;
+  /**
+   * image of the custom element to be saved
+   */
+  svg: string;
+
+  cutoff: number;
+  radius: number;
+  isFileInvalid: boolean;
+}
+
+export class CustomIconModal extends Component<Props, State> {
+  private _isMounted: boolean = false;
+
+  public state = {
+    symbolId: this.props.symbolId || '',
```

Comment:
Not sure symbolId is needed in state. It never gets set, other then here. How about removing symbolId from state and then just passing in `props.symbolId || getCustomIconId()` to onSave callback. Also, only show delete button if props.symbolId is provided